python/tank/bootstrap/manager.py

```diff
@@ -236,32 +254,126 @@ def bootstrap_engine(self, engine_name, entity=None):
+
+        return engine
+
+    def bootstrap_engine_async(self,
+                               engine_name,
+                               entity=None,
+                               completed_callback=None,
+                               failed_callback=None):
+        """
+        Create an :class:`~sgtk.Sgtk` instance for the given engine and entity,
+        then launch into the given engine.
+
+        This method launches the bootstrap process and returns immediately.
+        The :class:`~sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+        This will allow the main application to continue its execution and
+        remain responsive when bootstrapping the toolkit involves downloading files and
+        installing apps from the toolkit app store.
+
+        If entity is None, the method will bootstrap into the site
+        config. This method will attempt to resolve the config according
+        to business logic set in the associated resolver class and based
+        on this launch a configuration. This may involve downloading new
+        apps from the toolkit app store and installing files on disk.
+
+        2 callback functions can be provided.
+
+        A callback function that handles cleanup after successful completion of the bootstrap
+        with the following signature::
+
+            completed_callback(engine)
+
+        where:
+        - ``engine``is the launched :class:`~sgtk.platform.Engine` instance.
+
+        A callback function that handles cleanup after failed completion of the bootstrap
+        with the following signature::
+
+            failed_callback(phase, exception)
+
+        where:
+        - ``phase`` is the bootstrap phase that raised the exception,
+                    ``ToolkitManager.TOOLKIT_BOOTSTRAP_PHASE`` or ``ToolkitManager.ENGINE_STARTUP_PHASE``.
+                    Using this phase, the callback can decide if the toolkit core needs
+                    to be re-imported to ensure usage of a swapped in version.
+        - ``exception`` is the python exception raised while bootstrapping.
+
+        Please note that the API version of the tk instance that hosts
+        the engine may not be the same as the API version that was
+        executed during the bootstrap.
+
+        :param engine_name: Name of engine to launch (e.g. ``tk-nuke``).
+        :param entity: Shotgun entity to launch engine for.
+        :type entity: Dictionary with keys ``type`` and ``id``.
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        """
+
+        log.info("Bootstrapping engine %s for entity %s." % (engine_name, entity))
+
+        try:
+            import async
```

Comment:
further to josh's comments, can we rename this to `async_bootstrap`? I think that's clearer!